src/stream/src/executor/join/hash_join.rs

```diff
@@ -224,12 +231,160 @@ pub struct JoinHashMap<K: HashKey, S: StateStore> {
+/// If we obtain the degrees in a stream,
+/// we will need to hold an immutable reference to the state table for the entire lifetime,
+/// preventing us from concurrently updating the state table.
+///
+/// The cost of fetching all degrees upfront is acceptable. We currently already do so
+/// in `fetch_cached_state`.
+/// The memory use should be limited since we only store a u64.
+///
+/// Let's say we have amplification of 1B, we will have 1B * 8 bytes ~= 8GB
+///
+/// We can also have further optimization, to permit breaking the streaming update,
+/// to flush the in-memory degrees, if this is proven to have high memory consumption.
+///
+/// TODO(kwannoel): Perhaps we can cache these separately from matched rows too.
+/// Because matched rows may occupy a larger capacity.
+///
+/// Argument for this:
+/// We only hit this when cache miss. When cache miss, we will have this as one off cost.
+/// Keeping this cached separately from matched rows is beneficial.
+/// Then we can evict matched rows, without touching the degrees.
+async fn fetch_degrees<K: HashKey, S: StateStore>(
+    key: &K,
+    join_key_data_types: &[DataType],
+    degree_state_table: &StateTable<S>,
+) -> StreamExecutorResult<Vec<DegreeType>> {
+    let key = key.deserialize(join_key_data_types)?;
+    let mut degrees = vec![];
+    let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) = &(Bound::Unbounded, Bound::Unbounded);
+    let table_iter = degree_state_table
+        .iter_with_prefix(key, sub_range, PrefetchOptions::default())
+        .await
+        .unwrap();
+    #[for_await]
+    for entry in table_iter {
+        let degree_row = entry?;
+        let degree_i64 = degree_row
+            .datum_at(degree_row.len() - 1)
+            .expect("degree should not be NULL");
+        degrees.push(degree_i64.into_int64() as u64);
+    }
+    Ok(degrees)
+}
+
+// NOTE(kwannoel): This is not really specific to `TableInner`.
+// A degree table is `TableInner`, a `TableInner` might not be a degree table.
+// Hence we don't specify it in its impl block.
+pub(crate) fn update_degree<S: StateStore, const INCREMENT: bool>(
+    order_key_indices: &[usize],
+    degree_state: &mut TableInner<S>,
+    matched_row: &mut JoinRow<OwnedRow>,
+) {
+    let old_degree_row = matched_row
+        .row
+        .as_ref()
+        .project(order_key_indices)
+        .chain(once(Some(ScalarImpl::Int64(matched_row.degree as i64))));
+    if INCREMENT {
+        matched_row.degree += 1;
+    } else {
+        // DECREMENT
+        matched_row.degree -= 1;
+    }
+    let new_degree_row = matched_row
+        .row
+        .as_ref()
+        .project(order_key_indices)
+        .chain(once(Some(ScalarImpl::Int64(matched_row.degree as i64))));
+    degree_state.table.update(old_degree_row, new_degree_row);
+}
+
 pub struct TableInner<S: StateStore> {
     /// Indices of the (cache) pk in a state row
     pk_indices: Vec<usize>,
```

Comment:
Should this be `cache_key` instead? Or upstream pk indices?

Because `pk_indices` is too vague. It could refer to either:
1. `pk_indices` of the upstream table.
2. `pk_indices` of the join side table (includes join key prefix + upstream table pk).

Reply:
cc @yuhao-su

Reply:
yes `cache_key` should be a more accurate name. Thanks